File: backend/BeautyClinicSystem/src/main/java/com/project/service/ClientService.java
package com.project.service;

import com.project.dto.ClientDTO;
import com.project.mappers.ClientMapper;
import com.project.models.Client;
import com.project.repositories.ClientRepository;
import com.project.repositories.EstheticianRepository;
import com.project.security.AuthService;
import com.project.specification.ClientSpecification;
import lombok.RequiredArgsConstructor;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;

import java.time.OffsetDateTime;
import java.util.List;

import static org.springframework.http.HttpStatus.*;

@Service
@RequiredArgsConstructor
public class ClientService {

    private final ClientRepository clientRepository;
    private final AuthService authService;
    private final EstheticianRepository estheticianRepository;

    @Transactional(readOnly = true)
    public List<ClientDTO> listAll() {
        String loggedInEstheticianCpf = authService.getLoggedInEstheticianCpf();
        Specification<Client> spec = ClientSpecification.byEsthetician(loggedInEstheticianCpf);

        return clientRepository.findAll(spec).stream()
                .map(ClientMapper::fromEntityToDto)
                .toList();
    }

    @Transactional(readOnly = true)
    public List<ClientDTO> filterClients(String cpf, String name, OffsetDateTime startDate, OffsetDateTime endDate, Boolean isActive) {
        String loggedInEstheticianCpf = authService.getLoggedInEstheticianCpf();

        Specification<Client> spec = Specification
                .where(ClientSpecification.byEsthetician(loggedInEstheticianCpf))
                .and(ClientSpecification.cpfContains(cpf))
                .and(ClientSpecification.nameContains(name))
                .and(ClientSpecification.lastConsultationOn(startDate))
                .and(ClientSpecification.lastConsultationOn(endDate))
                .and(ClientSpecification.isActive(isActive));

        return clientRepository.findAll(spec).stream()
                .map(ClientMapper::fromEntityToDto)
                .toList();
    }

    @Transactional(readOnly = true)
    public ClientDTO findByCpf(String cpf) {
        Client client = clientRepository.findById(cpf)
                .filter(a -> a.getEsthetician().getCpf().equals(authService.getLoggedInEstheticianCpf()))
                .orElseThrow(() -> new ResponseStatusException(NOT_FOUND, "client-not-found"));
        return ClientMapper.fromEntityToDto(client);
    }

    @Transactional
    public ClientDTO create(ClientDTO createDTO) {
        if (clientRepository.existsById(createDTO.cpf())) {
            throw new ResponseStatusException(BAD_REQUEST, "cpf-already-registered");
        }

        Client entity = ClientMapper.fromDtoToEntity(createDTO);
        entity.setEsthetician(estheticianRepository.getReferenceById(authService.getLoggedInEstheticianCpf()));
        Client savedEntity = clientRepository.save(entity);
        return ClientMapper.fromEntityToDto(savedEntity);
    }

    @Transactional
    public ClientDTO update(String cpf, ClientDTO updateDTO) {
        String loggedInEstheticianCpf = authService.getLoggedInEstheticianCpf();
        Client entity = clientRepository.findById(cpf)
                .orElseThrow(() -> new ResponseStatusException(NOT_FOUND, "client-not-found"));

        if (!entity.getEsthetician().getCpf().equals(loggedInEstheticianCpf)) {
            throw new ResponseStatusException(FORBIDDEN, "cannot-update-client-for-another-esthetician");
        }

        entity.setName(updateDTO.name());
        entity.setPhone(updateDTO.phone());
        entity.setAddress(updateDTO.address());
        entity.setActive(updateDTO.isActive());
        entity.setLastConsultationDate(updateDTO.lastConsultationDate());

        Client updatedEntity = clientRepository.save(entity);
        return ClientMapper.fromEntityToDto(updatedEntity);
    }

    @Transactional
    public void delete(String cpf) {
        String loggedInEstheticianCpf = authService.getLoggedInEstheticianCpf();
        Client client = clientRepository.findById(cpf)
                .orElseThrow(() -> new ResponseStatusException(NOT_FOUND, "client-not-found"));

        if (!client.getEsthetician().getCpf().equals(loggedInEstheticianCpf)) {
            throw new ResponseStatusException(FORBIDDEN, "cannot-delete-client-for-another-esthetician");
        }

        clientRepository.deleteById(cpf);
    }
}

File: frontend/src/services/clientService.ts
import api from '../api/api';
import { ClientDTO } from '../types';

export const clientService = {
  // Lista todos os clientes
  listAll: async (): Promise<ClientDTO[]> => {
    const response = await api.get('/clients');
    return response.data;
  },

  // Busca cliente por CPF
  findByCpf: async (cpf: string): Promise<ClientDTO> => {
    const response = await api.get(`/clients/${cpf}`);
    return response.data;
  },

  // Cria novo cliente
  create: async (client: ClientDTO): Promise<ClientDTO> => {
    const response = await api.post('/clients', client);
    return response.data;
  },

  // Atualiza cliente
  update: async (cpf: string, client: ClientDTO): Promise<ClientDTO> => {
    const response = await api.put(`/clients/${cpf}`, client);
    return response.data;
  },

  // Deleta cliente
  delete: async (cpf: string): Promise<void> => {
    await api.delete(`/clients/${cpf}`);
  },

  filterClients: async (
    cpf?: string,
    name?: string,
    startDate?: string, 
    endDate?: string,
    isActive?: boolean
  ): Promise<ClientDTO[]> => {
    const params: any = {};
    if (cpf) params.cpf = cpf;
    if (name) params.name = name;
    if (startDate) params.startDate = startDate;
    if (endDate) params.endDate = endDate;
    if (isActive !== undefined) params.isActive = isActive;

    const response = await api.get('/clients/filter', { params });
    return response.data;
  }
};
